scripts/relup-base-vsns.escript

```diff
@@ -0,0 +1,278 @@
+upgrade/downgrade.
+
+We store a \"database\" of released versions as an `eterm' file, which
+is a mapping from a given version `Vsn' to its OTP version and a list
+of previous versions from which one can upgrade to `Vsn' (the
+\"from_versions\" list).  That allow us to more easily/explicitly keep
+track of allowed version upgrades/downgrades, as well as OTP changes
+between releases
+
+In the examples below, `VERSION_DB_PATH' represents the path to the
+`eterm' file containing the version database to be used.
+
+Usage:
+
+  * List the previous base versions from which `TO_VSN' may be
+    upgraded to.  Used to list versions for which relup files are to
+    be made.
+
+    relup-base-vsns.escript base-vsns TO_VSN VERSION_DB_PATH
+
+
+  * Show the OTP version with which `Vsn' was built.
+
+    relup-base-vsns.escript otp-vsn-for VSN VERSION_DB_PATH
+
+
+  * Automatically inserts a new version into the database.  Previous
+    versions with the same Major and Minor numbers as `Vsn' are
+    considered to be upgradeable from, and versions with higher Major
+    and Minor numbers will automatically include `Vsn' in their
+    \"from_versions\" list.
+
+    For example, if inserting 4.4.8 when 4.5.0 and 4.5.1 exists,
+    versions `BASE_FROM_VSN'...4.4.7 will be considered 4.4.8's
+    \"from_versions\", and 4.4.8 will be included into 4.5.0 and
+    4.5.1's from versions.
+
+    relup-base-vsns.escript insert-new-vsn NEW_VSN BASE_FROM_VSN OTP_VSN VERSION_DB_PATH
+
+  * Check if the version database is consistent considering `VSN'.
+
+    relup-base-vsns.escript check-vsn-db VSN VERSION_DB_PATH
+".
+
+main(["base-vsns", To0, VsnDB]) ->
+    VsnMap = read_db(VsnDB),
+    To = strip_pre_release(To0),
+    #{from_versions := Froms} = fetch_version(To, VsnMap),
+    AvailableVersionsIndex = available_versions_index(),
+    lists:foreach(
+     fun(From) ->
+       io:format(user, "~s~n", [From])
+     end,
+     filter_froms(Froms, AvailableVersionsIndex)),
+    halt(0);
+main(["otp-vsn-for", Vsn0, VsnDB]) ->
+    VsnMap = read_db(VsnDB),
+    Vsn = strip_pre_release(Vsn0),
+    #{otp := OtpVsn} = fetch_version(Vsn, VsnMap),
+    io:format(user, "~s~n", [OtpVsn]),
+    halt(0);
+main(["insert-new-vsn", NewVsn0, BaseFromVsn0, OtpVsn0, VsnDB]) ->
+    VsnMap = read_db(VsnDB),
+    NewVsn = strip_pre_release(NewVsn0),
+    validate_version(NewVsn),
+    BaseFromVsn = strip_pre_release(BaseFromVsn0),
+    validate_version(BaseFromVsn),
+    OtpVsn = list_to_binary(OtpVsn0),
+    case VsnMap of
+      #{NewVsn := _} ->
+          print_warning("Version ~s already in DB!~n", [NewVsn]),
+          halt(1);
+      #{BaseFromVsn := _} ->
+          ok;
+      _ ->
+          print_warning("Version ~s not found in DB!~n", [BaseFromVsn]),
+          halt(1)
+    end,
+    NewVsnMap = insert_new_vsn(VsnMap, NewVsn, OtpVsn, BaseFromVsn),
+    NewVsnList =
+        lists:sort(
+         fun({Vsn1, _}, {Vsn2, _}) ->
+           Vsn1 < Vsn2
```

Comment:
should compare the parsed result? otherwise `"4.4.8" > "4.4.10"` ?

Reply:
True, good catch! Thanks!